### PR TITLE
Add ADR proposal for handling stopped app routing

### PR DESCRIPTION
**Pull request by christianang**
### Summary of changes

Add an ADR proposal for how to handle routing to stopped applications. No decision has been made yet. We are opening this PR for comments on the best way forward to solve this problem.

Link to [ADR](https://github.com/cloudfoundry/cf-k8s-networking/blob/fb05a3dd692ab760373b783d3f1b6e1aaa772181/doc/architecture-decisions/0012-stopped-app-routing.md)

### Related Issue(s)
https://github.com/cloudfoundry/cf-for-k8s/issues/158

### Additional Context
https://www.pivotaltracker.com/story/show/172504850

@cloudfoundry/cf-networking 
@cloudfoundry/cf-capi

### CODE REVIEW

File: doc/architecture-decisions/0012-stopped-app-routing.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+## Proposal
+
+We have two proposals:
+
+1. Cloud Controller only adds destinations to Routes that are running. This way
+   Route objects remain the single source of truth for routing in the platform.
+   * _Note:_ Routes with _no_ destinations are explicitly supported already.
+
+2. Cloud Controller writes an App CR (this is a new CRD) that contains the
```
Comment (jrussett) [inline doc/architecture-decisions/0012-stopped-app-routing.md]: This second proposal seems like a whole lot of implementation on CAPI's end to figure out if the application instance is started or not. I feel like while both implementations are mutually exclusive, the first proposal can be an evolutionary step to the second proposal, at least until there are more use cases for holding all of the application information on an App CR.

Comment (XanderStrike) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: Totally agree. I think an App CR is an eventual goal for CAPI anyway, so this could be some evidence for prioritization of that. If they were planning to do it sooner rather than later we could tack this onto that effort.

Personally I strongly prefer proposal 1 because it's the least work for everyone, and the contents of Route are always true.

File: doc/architecture-decisions/0012-stopped-app-routing.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+currently know about the status of apps. They assume all destinations on a Route
+object are up and able to receive traffic.
+
+See this issue: https://github.com/cloudfoundry/cf-for-k8s/issues/158
+
+## Proposal
+
+We have two proposals:
+
+1. Cloud Controller only adds destinations to Routes that are running. This way
```
Comment (selzoc) [inline doc/architecture-decisions/0012-stopped-app-routing.md]: I think there's a distinction to be made here between apps that are stopped intentionally, i.e. `cf stop`, and apps that are crashing or haven't come up yet or otherwise unhealthy.  Which case is this proposal trying to address?

Comment (XanderStrike) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: Stopped intentionally. I would expect a crashed or not-yet-ready app to return 503s.

Comment (selzoc) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: Ah, great.  Then we can rely on CAPI's state and not have to reach out to k8s or eirini to figure it out 👍

Comment (tcdowney) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: Could just update this thing and this would be implemented. 😅 
https://github.com/cloudfoundry/cloud_controller_ng/blob/9461ff92e68382b541c76ff27f8111eb2f111cb6/lib/cloud_controller/backends/copilot_runner_delegator.rb#L2

Comment (tcdowney) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: One interesting thing about this, though, is that the Route CRs in k8s will no longer match as closely with what `/v3/routes` return... that's probably ok. I think I just change my mental model of what "desired vs actual state" means for routes.

Comment (rosenhouse) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: > I would expect a crashed or not-yet-ready app to return 503s.

I'd expect requests to a *route* to return 503 only if all of its destinations are stopped or crashed.  If any of them are healthy, then I'd expect requests to that route to succeed.  That is the `gorouter` behavior described in the Context section above.

Comment (rosenhouse) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: > the Route CRs in k8s will no longer match as closely with what`/v3/routes` return... that's probably ok

This would make me sad.  I'd really like Route to be a high-level description of desired intent, not an intermediate representation of the current started/stopped state of the destinations.

File: doc/architecture-decisions/0012-stopped-app-routing.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+   Route objects remain the single source of truth for routing in the platform.
+   * _Note:_ Routes with _no_ destinations are explicitly supported already.
+
+2. Cloud Controller writes an App CR (this is a new CRD) that contains the
+   application's status, so that the routecontroller can use that to decide how
+   to configure Istio.
+
+## Alternatives Considered
```
Comment (rosenhouse) [inline doc/architecture-decisions/0012-stopped-app-routing.md]: What about using Kubernetes [Readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes)?

From the docs:
> A pod with containers reporting that they are not ready does not receive traffic through Kubernetes Services.

Comment (XanderStrike) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: When an app is stopped in CF, the pod and statefulset are deleted, so we can't probe them for readiness.

Services which have selectors which are not fulfilled by any pods result in 503s. We have one service per CF app. So when an app is stopped the service for that app results in a 503.

One alternative we thought about that takes advantage of that aspect of services, but thought was too ridiculous to include was:
* Have capi label every pod with every route that maps to it
* Make one service per route that selects that label
* Make one virtualservice per route that selects the service

This doesn't solve for the situation when _no apps are started_, but it would prevent routing to apps that are down. It doesn't strike me as healthy to include routing information in the labels of the pods though, and it seemed like it'd be a lot of work for capi. You also lose weighted routing (and possibly some other features, haven't really checked) with this approach.

File: doc/architecture-decisions/0012-stopped-app-routing.md (hunk 2 above)
Comment (XanderStrike) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: > I'd really like Route to be a high-level description of desired intent

In that case, does proposal 2 work for you? That way the Route resource expresses routing intent to App resources which exist regardless of the Pods, instead of expressing a desire to route to Pods that may or may not exist. (imo, a desire to route to a backend that doesn't exist is most correctly served by a 503 error)

Comment (tcdowney) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: Yeah, that's what I thought as well... but my "probably ok" was thinking in terms of hitting GA targets for this. I think doing this optional wouldn't preclude us from moving to a more pure high-level description of desired intent later (like with the `App` CRD approach and what @jrussett is getting at in [this comment](https://github.com/cloudfoundry/cf-k8s-networking/pull/46/files#r432754776)).

As an aside, I was surprised to find out Istio was behaving this way... I thought it would behave more like how `Services` do when a subset of `Endpoints` are no longer healthy and just not route to them. I had imagined the `VirtualService` would understand that some of its backends were unhealthy and ignore them... even if we did declare some percentage of traffic should go to those. After this behavior was described to me what Istio is doing sounds like a valid approach -- just suprising to me, though, based on how other things in Kubernetes work.

File: doc/architecture-decisions/0012-stopped-app-routing.md (hunk 3 above)
Comment (cwlbraa) [reply doc/architecture-decisions/0012-stopped-app-routing.md:36]: If the app is stopped, there will be no eirini statefulset and therefore no pods to put readiness probes on.

File: doc/architecture-decisions/0012-stopped-app-routing.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+   causes a lot of no-op reconcile events for the same reason as above. It also
+   doesn't feel good to attach finalizers to every pod.
+3. Configure envoy to perform retries with circuit breaking to cull stopped
+   backends. This has many problems including: hiding real errors, increasing
+   latency when backends come back up, etc.
+
+After discussion, we feel that these alternatives constitute a workaround for
+invalid configuration (Routes referring to destinations that do not exist).
+
+Proposal 2 is also a type of workaround for invalid configuration, but we feel
```
Comment (rosenhouse) [inline doc/architecture-decisions/0012-stopped-app-routing.md]: I think it is totally valid for a CF Route to have a stopped App as one of its destinations.  That behavior has worked for years, and is likely baked into many blue/green deploy workflows used by CF users.

**Comment by rosenhouse** [comment]
Thanks all.  This one is tricky, and I learned a lot reading through this discussion.

I'm supportive of the idea of starting with Proposal 1 (where the Route is more of an intermediate representation of app state), and iterating towards a model where it is fully declarative of high-level intent, once App status is easier to discover from the k8s api.

File: doc/architecture-decisions/0012-stopped-app-routing.md (hunk 3 above)
Comment (rosenhouse) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: Makes sense.  May be worth summarizing this point as an "Alternative considered" so that future team members don't ask the same question that I did.

**Comment by rosenhouse** [comment]
Though we said crashed apps are out of scope here, won't we be facing a similar regression there as well?

Suppose someone has Route R with destination apps A and B.  A is healthy.  B is a broken build, and all instance crash on startup.

- In cf-deployment, instances of B crash, are deregistered over NATS, so GoRouter doesn't even try to connect to them.  End-user is only ever routed to A.

- In CF-K8s-Networking, instances of B crash, but Istio still treats the K8s Service as a destination and will load balance requests to it.

File: doc/architecture-decisions/0012-stopped-app-routing.md (hunk 4 above)
Comment (shalako) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: +1

Comment (christianang) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: Agreed, I think my mental model was flawed in considering the Route CR to be more closely aligned to actual state rather than the desired state. However, I think having more of the application state in Kubernetes will make this feel more correct because the App GUID will actually refer to something rather than it potentially referring to StatefulSets and Pods that potentially disappear when the app is stopped.

**Comment by christianang** [comment]
> Though we said crashed apps are out of scope here, won't we be facing a similar regression there as well?

Possibly, I've been assuming that crashed pods react the same way as a non-existent pod w.r.t a service. However, reading about readiness probes makes me believe it may act differently. I think it is probably worth some experimenting to see if they do act similarly.

**Comment by tcdowney** [comment]
One thing for any @cloudfoundry/cf-capi folk reviewing this: I think if we go through with **Proposal 1** it might be worth revisiting the decision to defer the CCDB->k8s `Route` bulk sync work. 

Proposal 1 will require Cloud Controller to update Kubernetes `Route` resources more often -- whenever a Cloud Controller `Process` has a state transition. Previously they would have only been updated on calls that explicitly dealt with Cloud Controller `Route`s (e.g. `cf create-route`, `cf delete-route`, `cf map-route`, `cf unmap-route`, and server side manifests). Now we have to consider `cf restart`, `cf stop`, `cf start`, `cf push`, etc. since they all tweak `Process` state.

Also, since the obvious way of implementing this on Cloud Controller involves the [`ProcessObserver`](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/lib/cloud_controller/process_observer.rb) which gets invoked in ["After Commit"](https://github.com/cloudfoundry/cloud_controller_ng/blob/6874526615a4645a555123052d74054ffa36610e/app/models/runtime/process_model.rb#L543) hooks I think it's more likely we'll see cases where CCDB updates have been made, but the Kubernetes update has failed. A lot of the implementation decisions in that part of the Cloud Controller code were made with the assumption that the Cloud Controller/Diego bulk sync would make things eventually consistent.

**tl;dr**
I think we're going to encounter more Route consistency issues and will need the bulk sync.

File: doc/architecture-decisions/0013-stopped-app-routing.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+healthy app instances]
+
+This is because the components that create the Istio configuration do not
+currently know about the status of apps. They assume all destinations on a Route
+object are up and able to receive traffic. More concretely the Services that are
+created use the app guid and process type as selectors. When an application is
+stopped the StatefulSet and Pods backing that application are deleted. This
+means the Selector can't select any pods and results in a 503.
+
+This ADR is specifically focused on handling routing to stopped applications.
```
Comment (jenspinney) [inline doc/architecture-decisions/0013-stopped-app-routing.md:35]: I'm wondering if it should be the other way around: we first dive into and focus on how to handle the crashed app situation, since that could be viewed as a more extreme version of the stopped app case. If we determine how we plan to avoid sending traffic to crashed app instances, then dealing with the stopped app case is likely pretty simple at that point.

**Comment by christianang** [comment]
> Though we said crashed apps are out of scope here, won't we be facing a similar regression there as well?

Back to this question. I ran a couple experiments to understand how crashed apps are handled at the moment.

Scenario: Pushed two apps mapped to the same route. Crashed one app.
Result: Intermittent 503s
Reason: The Virtual Service results in two Services as destinations and the Virtual Service balances traffic between them even if one Service results in 503s.

Scenario: Push single app with one route and multiple instances. Crash one app instance.
Result: No 503s. Only 200s.
Reason: The Virtual Service results in one Service as the destination and can take advantage of Readiness Probes on the pod.

So this is also a problem that will need addressing, that wouldn't be solved by anything proposed so far here. Perhaps routing to crashed apps and stopped apps can be solved in the same way, but not sure at the moment. However, I do agree with @jenspinney that it is probably worth solving crashed apps first as solution may solve or fit into how to solve routing to stopped apps.

File: doc/architecture-decisions/0012-stopped-app-routing.md (hunk 2 above)
Comment (cwlbraa) [reply doc/architecture-decisions/0012-stopped-app-routing.md]: Can someone convince me the same way @tcdowney 's been convinced that `Route`s and `VirtualService`s should continue trying to route to backends that don't exist? 

> a desire to route to a backend that doesn't exist is most correctly served by a 503 error

But when 1/3 backends don't exist, a desire to route to 3 backends is most correctly served by a 503 error?

**Comment by cwlbraa** [comment]
The more I read the more I'm convinced that Proposal 2 is potentially not only cleaner, but also easier.

**Comment by XanderStrike** [comment]
We're gonna call this blocked on [#173159123](https://www.pivotaltracker.com/story/show/173159123)

**Comment by XanderStrike** [comment]
We're gonna go to GA with https://github.com/cloudfoundry/cf-for-k8s/issues/158 as a known issue and keep our ears open for feedback from users on if/how we want this handled.

**Comment by loewenstein** [comment]
Did we consider using Envoy's health checking mechanism? There is an [open issue for Istio](https://github.com/istio/istio/issues/10537) that has [some solution proposed](https://github.com/istio/istio/issues/10537#issuecomment-555548652) although most of the issue focuses on external services.

BTW why does Istio passive health checking logic not help in this case? I'd expect not routing traffic to pods that are not healthy or not even existing would be covered by default.

**Comment by XanderStrike** [comment]
Envoy's health checking can reject endpoints in the `cluster` for being unhealthy, it cannot reject an entire `cluster` and redistribute traffic in a `weighted_clusters` set. Envoy choses the cluster before checking which endpoints in the cluster are available. For this to work for us we'd have to contribute to Envoy to flip that process, and I'm not sure they'd go for it. From their perspective, asking Envoy to route to a cluster with no available endpoints is more of a misconfiguration than something the proxy should be expected to solve.

**Comment by KauzClay** [comment]
We are going to close this PR as we believe the solution is outdated. Following comments in this [story](https://www.pivotaltracker.com/n/projects/2407973/stories/173159123), we believe this isn't an immediate priority either. 

We can either make a new PR or reopen this one when we get to it in the future.
